github/oidctest.go

```diff
@@ -1,41 +1,130 @@
+	}
+	return payload, nil
+}
+
 // NewTestOIDCServer returns an httptest.Server that can be used as the OIDC
-// server in tests and a cleanup function that can be used to stop and clean up
-// the server. The server returns the given token when queried.
-func NewTestOIDCServer(t *OIDCToken) (*httptest.Server, func()) {
-	b, err := json.Marshal(t)
+// server, and an OIDClient that will use the test server. The server returns the
+// given token when queried. Now is the time used for token expiration
+// verification by the client.
+func NewTestOIDCServer(t *testing.T, now time.Time, token *OIDCToken) (*httptest.Server, *OIDCClient) {
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	signer, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.RS256, Key: privateKey}, nil)
 	if err != nil {
-		panic(err)
+		t.Fatal(err)
 	}
 
-	rawResponse := fmt.Sprintf(`{"value": "part1.%s.part3"}`, base64.RawURLEncoding.EncodeToString(b))
-	return newRawTestOIDCServer(rawResponse)
+	// FIXME: Fix creating a test server that can return tokens that can be verified.
+	var issuerURL string
+	s, c := newTestOIDCServer(t, now, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		b, err := json.Marshal(jsonToken{
+			Issuer:         issuerURL,
+			Audience:       r.URL.Query()["audience"],
+			Expiry:         token.Expiry.Unix(),
+			JobWorkflowRef: token.JobWorkflowRef,
+		})
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		object, err := signer.Sign(b)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		value, err := object.CompactSerialize()
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		fmt.Fprintf(w, `{"value": "%s"}`, value)
+	}))
+	issuerURL = s.URL
+
+	return s, c
 }
 
-func newRawTestOIDCServer(raw string) (*httptest.Server, func()) {
-	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+func newRawTestOIDCServer(t *testing.T, now time.Time, status int, raw string) (*httptest.Server, *OIDCClient) {
+	return newTestOIDCServer(t, now, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		// Respond with a very basic 3-part JWT token.
+		w.WriteHeader(status)
 		fmt.Fprintln(w, raw)
 	}))
-	oldEnv, ok := os.LookupEnv(requestURLEnvKey)
-	// NOTE: httptest.Server.URL has no trailing slash.
-	os.Setenv(requestURLEnvKey, s.URL+"/")
-	return s, func() {
-		s.Close()
-		if ok {
-			os.Setenv(requestURLEnvKey, oldEnv)
-		} else {
-			os.Unsetenv(requestURLEnvKey)
+}
+
+func newTestOIDCServer(t *testing.T, now time.Time, f http.HandlerFunc) (*httptest.Server, *OIDCClient) {
+	var issuerURL string
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case "/":
+			f(w, r)
+		case "/.well-known/openid-configuration":
```

Comment:
This path isn't actually used since the test oidc client's verifierFunc doesn't instantiate a provider.

Reply:
Will leave it in case `oidc.Provider` is used by library consumers.